### PR TITLE
vptranslator: rewrite

### PR DESCRIPTION
**Pull request by TeXitoi**
Rewrite vptranslator (the thing that transform a ValidityPattern in human readable form).

### CODE REVIEW

**Comment by TeXitoi** [comment]
See https://github.com/TeXitoi/pinot2015presenter for a (work in progress in french) description of the algorithm

**Comment by TeXitoi** [comment]
jenkins, test this please.

File: source/vptranslator/vptranslator.cpp [hunk 1]
```diff
@@ -29,321 +29,236 @@ www.navitia.io
...
-    //tant que le prochain lundi est inférieur à enddate, on découpe le régime
-        if(pos < CS.length()){
-            substr = CS.substr(precpos, pos - precpos);
+void BlockPattern::canonize_validity_period() {
+    if (validity_period.empty()) return;
+    std::set<boost::gregorian::date_period> res;
+    auto it = validity_period.cbegin(), end = validity_period.cend();
+    auto cur_period = *it;
+    for (++it; it != end; ++it) {
+        if (cur_period.end() == it->begin()) {
```
Comment (l-vincent-l) [inline source/vptranslator/vptranslator.cpp:90]: a comment would be nice

File: source/vptranslator/vptranslator.cpp [hunk 2]
```diff
@@ -29,321 +29,236 @@ www.navitia.io
...
+    }
+    res.insert(cur_period);
+    validity_period = res;
+}
 
-        std::map<int, target>::iterator it = target_map.find(currentweek.week_bs.to_ulong());
-        if (it != target_map.end()){
-            ++it->second.count;
-            it->second.lastweeknumber = weeknumber;
+unsigned BlockPattern::nb_weeks() const {
```
Comment (l-vincent-l) [inline source/vptranslator/vptranslator.cpp:101]: can you explain what you're doing?

File: source/vptranslator/vptranslator.cpp [hunk 3]
```diff
@@ -29,321 +29,236 @@ www.navitia.io
...
-    std::bitset<7> result = dest.flip() &= b2;
-    return result;
+void BlockPattern::add_from(const BlockPattern& other) {
+    if (!other.excluding.empty() || !other.excluding.empty()) {
+        throw std::invalid_argument("add_from takes only no exception block patterns");
+    }
+    const Week diff = week ^ other.week;
+    const Week to_exclude = ~other.week & diff;
+    const Week to_include = other.week & diff;
+    for (const auto& vp: other.validity_period) {
```
Comment (l-vincent-l) [inline source/vptranslator/vptranslator.cpp]: what is vp? a period? If so can you name the variable period?

File: source/vptranslator/vptranslator.cpp [hunk 4]
```diff
@@ -29,321 +29,236 @@ www.navitia.io
...
+        validity_period.insert(vp);
+        for (day_iterator it = vp.begin(); it != vp.end(); ++it) {
```
Comment (l-vincent-l) [inline source/vptranslator/vptranslator.cpp]: Can you name it day_it?

File: source/vptranslator/vptranslator.cpp [hunk 5]
```diff
@@ -29,321 +29,236 @@ www.navitia.io
...
-
-            if(exception.any()){
-                for(int it=6; it!= 6 + weeklength; it--) {
-                    if(exception[it]){
-                        responselw.exceptlist.push_back(endweek.startdate + boost::gregorian::date_duration(7 - it -1));
-                    }
-                }
+namespace {
+struct DistanceMatrix: boost::noncopyable {
+    static const unsigned size = 1 << 7;
```
Comment (l-vincent-l) [inline source/vptranslator/vptranslator.cpp:163]: A comment would be nice.

File: source/vptranslator/vptranslator.cpp [hunk 6]
```diff
@@ -29,321 +29,236 @@ www.navitia.io
...
-        //on cree un vector de pointeur sur une copie des elements du target_map
-        std::vector<target> targetvector;
-        for(std::map<int, target>::iterator it=target_map.begin(); it!= target_map.end(); ++it) {
-            targetvector.push_back(it->second);
+std::vector<BlockPattern> translate_no_exception(const ValidityPattern& vp) {
+    unsigned offset = 0, end = end_active(vp);
+    std::unordered_map<Week, BlockPattern> patterns;
+    
+    // skipping zeros at the beginning
+    for (offset = 0; offset < end && !vp.check(offset); ++offset) {}
```
Comment (l-vincent-l) [inline source/vptranslator/vptranslator.cpp:236]: you could use the find_first function

File: source/vptranslator/vptranslator.cpp (hunk 2 above)
Comment (TeXitoi) [inline source/vptranslator/vptranslator.cpp:101]: See the header

File: source/vptranslator/vptranslator.cpp (hunk 6 above)
Comment (TeXitoi) [inline source/vptranslator/vptranslator.cpp:236]: I need the index, not the iterator

File: source/vptranslator/vptranslator.cpp (hunk 1 above)
Comment (TeXitoi) [inline source/vptranslator/vptranslator.cpp:90]: Even after reading the comment in the header?

File: source/vptranslator/vptranslator.cpp (hunk 2 above)
Comment (l-vincent-l) [inline source/vptranslator/vptranslator.cpp:101]: My comment was confusing, can you explain how you're doing it.
for instance I don't get why: 
  if (idx_begin <= idx_end) {
     res += 2;

File: source/vptranslator/vptranslator.h [hunk 7]
```diff
@@ -31,90 +31,38 @@ www.navitia.io
...
+namespace navitia {
+namespace vptranslator {
 
-template<class T>
-unsigned int levenshtein_distance(const T &s1, const T & s2) {
-        const size_t len1 = s1.size(), len2 = s2.size();
-        std::vector<unsigned int> col(len2+1), prevCol(len2+1);
+struct BlockPattern {
+    navitia::type::Calendar::Week week;
+    std::set<boost::gregorian::date_period> validity_period;
```
Comment (l-vincent-l) [inline source/vptranslator/vptranslator.h]: can you rename it in validity_periods?

File: source/vptranslator/vptranslator.cpp (hunk 1 above)
Comment (l-vincent-l) [inline source/vptranslator/vptranslator.cpp:90]: Just comment why is it working?
(it's ordered, and every period has 7 days starting on the same day of the week, so there's no over-lapping)

File: source/vptranslator/vptranslator.cpp (hunk 2 above)
Comment (TeXitoi) [inline source/vptranslator/vptranslator.cpp:101]: Comments added.

**Comment by TeXitoi** [comment]
Sample output:

``` javascript
vehicle_journeys": [

    {
        "calendars": [
            {
                "active_periods": [
                    {
                        "begin": "20141103",
                        "end": "20141222"
                    },
                    {
                        "begin": "20150105",
                        "end": "20150131"
                    }
                ],
                "exceptions": [
                    {
                        "datetime": "20141111",
                        "type": "remove"
                    }
                ],
                "id": "",
                "name": "",
                "validity_pattern": {
                    "beginning_date": "",
                    "days": ""
                },
                "week_pattern": {
                    "friday": true,
                    "monday": true,
                    "saturday": false,
                    "sunday": false,
                    "thursday": true,
                    "tuesday": true,
                    "wednesday": false
                }
            }
        ],
        "comment": "",
        "id": "vehicle_journey:ART:104-4120355",
        "journey_pattern": {
            "id": "journey_pattern:ART:104-1188",
            "name": "Square Frachon"
        },
        "name": "ART:104-4120355",
        "stop_times": [
            {
                "arrival_time": "170000",
                "departure_time": "170000"
            },
            {
                "arrival_time": "171500",
                "departure_time": "171500"
            }
        ],
        "validity_pattern": {
            "beginning_date": "20141031",
            "days": "000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000011011001101100110110011011000000000000000011011001101100110110011011001101100110010011011000"
        }
    }
]
```

I can't find how to remove id, name and validity_pattern, they are not setted in the protobuf

**Comment by TeXitoi** [comment]
jenkins, test this please.

**Pull request merged**